base/abstractarray.jl

```diff
@@ -373,9 +373,16 @@ zero{T}(x::AbstractArray{T}) = fill!(similar(x), zero(T))
 # While the definitions for LinearFast are all simple enough to inline on their
 # own, LinearSlow's CartesianRange is more complicated and requires explicit
 # inlining.
-start(A::AbstractArray) = (@_inline_meta(); itr = eachindex(A); (itr, start(itr)))
-next(A::AbstractArray,i) = (@_inline_meta(); (idx, s) = next(i[1], i[2]); (A[idx], (i[1], s)))
-done(A::AbstractArray,i) = done(i[1], i[2])
+start(A::AbstractArray) = (@_inline_meta(); _start(linearindexing(A), A))
+_start(::LinearFast, A::AbstractArray) = 1
+_start(::LinearSlow, A::AbstractArray) = (@_inline_meta(); itr = eachindex(A); (itr, start(itr)))
```

Comment:
I was just wondering whether there was something weird in the strings code. In general, I agree one shouldn't call `start` and then assume the returned value is an integer. But when implementing a specific type, it's fine to rely on knowing that its `start` method returns a special index type which supports arithmetic, like in https://github.com/JuliaLang/julia/issues/9297 (which isn't implemented at all, so not a justification for the current code).

I think the reason that code is written like this is that [this method](https://github.com/JuliaLang/julia/blob/c18bc5383f6cc4d95cab70e7312a9eb15e0c5b95/base/unicode/checkstring.jl#L155) accepts both `Vector` and `AbstractString` objects. `start` could/should be replaced with `1` here, but `endof` is definitely needed (and we assume it returns an integer index). BTW, this code deals with UTF-16 and UTF-32 strings, so it's likely to go away soon.